File: Haskell/problem6.hs
{- (*) Find out whether a list is a palindrome.
       A palindrome can be read forward or backward; e.g. (x a m a x). -}

isPalindrome :: (Eq a) => [a] -> Bool
isPalindrome xs = xs == reverse xs

isPalindrome' :: (Eq a) => [a] -> Bool
isPalindrome' xs = isPalindrome_' xs xs []

isPalindrome_' :: (Eq a) => [a] -> [a] -> [a] -> Bool 
isPalindrome_' (x:xs) (_:_:ys) acc = isPalindrome_' xs ys (x:acc)
isPalindrome_' (_:xs) [_] acc = isPalindrome_' xs [] acc
isPalindrome_' xs [] acc = xs == acc

File: Prolog/Prolog_Lists/problem6.pl
%% 1.06 (*) Find out whether a list is a palindrome.

% palindrome(+List)
palindrome(List) :- palindrome(List, List, []).

palindrome(Xs, [], Xs) :- !. % green cut
palindrome([_|Xs], [_], Xs) :- !. % green cut
palindrome([X|Xs], [_,_|Ys], Acc) :- palindrome(Xs, Ys, [X|Acc]).
